pennylane_lightning/src/benchmarks/Bench_ApplyOperations.cpp

```diff
@@ -0,0 +1,196 @@
+BENCHMARK_CAPTURE(serializeOps, ops_all, {"PauliX",     "PauliY",
+                                          "PauliZ",     "Hadamard",
+                                          "S",          "T",
+                                          "RX",         "RY",
+                                          "RZ",         "Rot",
+                                          "PhaseShift", "CNOT",
+                                          "SWAP",       "ControlledPhaseShift",
+                                          "CRX",        "CRY",
+                                          "CRZ",        "CRot",
+                                          "Toffoli",    "CSWAP"});
+
+static inline auto generateNeighboringWires(size_t start_idx, size_t num_qubits,
+                                            size_t num_wires)
+    -> std::vector<size_t> {
+    std::vector<size_t> v;
+    v.reserve(num_wires);
+    for (size_t k = 0; k < num_wires; k++) {
+        v.emplace_back((start_idx + k) % num_qubits);
+    }
+    return v;
+}
+
+static void applyOperationsFromRandOps(benchmark::State &state,
+                                       Pennylane::Gates::KernelType kernel,
+                                       const std::vector<OpName> op_names) {
+    if (op_names.empty()) {
+        state.SkipWithError("Invalid list of operations.");
+    }
+
+    size_t num_gates = state.range(0);
+    size_t num_qubits = state.range(1);
+
+    if (!num_gates) {
+        state.SkipWithError("Invalid number of gates.");
+    }
+
+    if (!num_qubits) {
+        state.SkipWithError("Invalid number of qubits.");
+    }
+
+    auto ops = serializeOperationsFromStrings(state, op_names);
+
+    std::random_device rd;
+    std::mt19937_64 eng(rd());
+
+    std::uniform_int_distribution<size_t> gate_distr(0, ops.size() - 1);
+    std::uniform_int_distribution<size_t> inv_distr(0, 1);
+    std::uniform_real_distribution<double> param_distr(0.0, 2 * M_PI);
+    std::uniform_int_distribution<size_t> wire_distr(0, num_qubits - 1);
+
+    auto param_generator = [&param_distr, &eng]() { return param_distr(eng); };
+
+    std::vector<std::string_view> rand_gate_names;
+    std::vector<std::vector<size_t>> rand_gate_wires;
+    std::vector<std::vector<double>> rand_gate_params;
+
+    std::vector<bool> rand_inverses;
+
+    for (size_t i = 0; i < num_gates; i++) {
+        size_t wire_start_idx = wire_distr(eng);
+        const auto &[op_name, n_wires, n_params] = ops[gate_distr(eng)];
+
+        rand_gate_names.emplace_back(op_name);
+        rand_gate_wires.emplace_back(
+            generateNeighboringWires(wire_start_idx, num_qubits, n_wires));
+
+        std::vector<double> params(n_params);
+        std::generate(params.begin(), params.end(), param_generator);
+        rand_gate_params.emplace_back(std::move(params));
+
+        rand_inverses.emplace_back(static_cast<bool>(inv_distr(eng)));
+    }
+
+    for (auto _ : state) {
+        Pennylane::StateVectorManaged<double> sv{num_qubits};
```

Comment:
Again, just wondering if these can be templated for different precisions?

Reply:
Oops! Thanks for catching this. Done.